File: src/server/MessageFactory.cs
using System;
using System.Text;
using Newtonsoft.Json;


namespace WebTanksServer
{
    internal static class MessageFactory
    {
        public static MessageBase DeserializeMessage(ArraySegment<byte> message)
        {
            Message deserializedMessage;
            dynamic content;
            try
            {
                deserializedMessage = JsonConvert.DeserializeObject<Message>(Encoding.UTF8.GetString(message));
                content = JsonConvert.DeserializeObject(deserializedMessage.Content);
            }
            catch
            {
                return null;
            }
            switch (deserializedMessage.Type)
            {
                case MessageType.SetName:
                    return CreateMessageSetName(content);
                case MessageType.SetMap:
                    return CreateMessageSetMap(content);
                case MessageType.StartGame:
                    return CreateMessageStartGame(content);
                case MessageType.PressedButton:
                    return CreateMessagePressedButton(content);
                case MessageType.EndGame:
                    return CreateMessageEndGame(content);
                default:
                    return null;
            }
        }

        public static string SerealizeMessage(dynamic content)
        {
            var message = new Message();
            message.Content = JsonConvert.SerializeObject(content);
            switch (content.Type)
            {
                case MessageType.SetName:
                    message.Type = MessageType.SetName;
                    return JsonConvert.SerializeObject(message);
                case MessageType.SetMap:
                    message.Type = MessageType.SetMap;
                    return JsonConvert.SerializeObject(message);
                case MessageType.StartGame:
                    message.Type = MessageType.SetName;
                    return JsonConvert.SerializeObject(message);
                case MessageType.PressedButton:
                    message.Type = MessageType.SetName;
                    return JsonConvert.SerializeObject(message);
                case MessageType.EndGame:
                    message.Type = MessageType.SetName;
                    return JsonConvert.SerializeObject(message);
                default:
                    return null;
            }
        }

        private static MessageSetName CreateMessageSetName(dynamic content)
        {
            MessageSetName setName = new MessageSetName();
            setName.Name = content.Name;
            return setName;
        }

        private static MessageSetMap CreateMessageSetMap(dynamic content)
        {
            MessageSetMap setMap = new MessageSetMap();
            setMap.Name = content.Name;
            setMap.Map = content.Map;
            return setMap;
        }

        private static MessageStartGame CreateMessageStartGame(dynamic content)
        {
            MessageStartGame startGame = new MessageStartGame();
            startGame.Name = content.Name;
            return startGame;
        }

        private static MessagePressedButton CreateMessagePressedButton(dynamic content)
        {
            MessagePressedButton pressedButton = new MessagePressedButton();
            pressedButton.Button = content.PressedButton;
            pressedButton.Name = content.Name;
            return pressedButton;
        }

        private static MessageEndGame CreateMessageEndGame(dynamic content)
        {
            MessageEndGame endGame = new MessageEndGame();
            endGame.Name = content.Name;
            return endGame;
        }

        private static string DeleteUslessSymbols(string content)
        {
            char[] symbols = content.ToCharArray();
            int breacketsCounter = 0;
            for (int i = 0; i < symbols.Length; i++)
            {
                switch (symbols[i])
                {
                    case '{':
                        breacketsCounter++;
                        break;
                    case '}':
                        breacketsCounter--;
                        if (breacketsCounter == 0)
                        {
                            return content.Substring(0, i);
                        }
                        break;
                }
            }
            return null;
        }
    }
}

File: src/client/ts/message-factory.ts
import { MessageSetName } from './models/message-models/message-set-name';
import { MessageType } from './models/message-models/message-type';
import { Message } from './models/message-models/message';
import { MessageSetMap } from './models/message-models/message-set-map';
import { MessageStartGame } from './models/message-models/message-start-game';
import { MessagePressedButton } from './models/message-models/message-pressed-button';
import { KeyCode } from './models/key-code';
import { MessageEndGame } from './models/message-models/message-end-game';

export class MessageFactory {
    public createMessageSetName(name: string): string {
        let message = new MessageSetName();
        message.Type = MessageType.SetName;
        message.Name = name;
        let stringMessage = JSON.stringify(message);
        let readyMessage = new Message();
        readyMessage.Type = MessageType.SetName;
        readyMessage.Content = stringMessage;
        return JSON.stringify(readyMessage);
    }

    public createMessageSetMap(map: string, name: string): string {
        let message = new MessageSetMap();
        message.Type = MessageType.SetMap;
        message.Map = map;
        message.Name = name;
        let stringMessage = JSON.stringify(message);
        return this.createMainMessage(stringMessage, message.Type);
    }

    public createMessageStartGame(name: string): string {
        let message = new MessageStartGame();
        message.Type = MessageType.StartGame;
        message.Name = name;
        let stringMessage = JSON.stringify(message);
        let readyMessage = new Message();
        readyMessage.Type = MessageType.StartGame;
        readyMessage.Content = stringMessage;
        return this.createMainMessage(stringMessage, message.Type);
    }

    public createMessagePressedButton(button: KeyCode, name: string): string {
        let message = new MessagePressedButton();
        message.Type = MessageType.PressedButton;
        message.Button = button;
        message.Name = name;
        let stringMessage = JSON.stringify(message);
        let readyMessage = new Message();
        readyMessage.Type = MessageType.PressedButton;
        readyMessage.Content = stringMessage;
        return this.createMainMessage(stringMessage, message.Type);
    }

    public createMessageEndGame(name: string): string {
        let message = new MessageEndGame();
        message.Type = MessageType.EndGame;
        message.Name = name;
        let stringMessage = JSON.stringify(message);
        let readyMessage = new Message();
        readyMessage.Type = MessageType.EndGame;
        readyMessage.Content = stringMessage;
        return this.createMainMessage(stringMessage, message.Type);
    }

    private createMainMessage(message: string, type: MessageType): string {
        let readyMessage = new Message();
        readyMessage.Type = type;
        readyMessage.Content = message;
        return JSON.stringify(readyMessage);
    }
}